### ISSUE TITLE
🐛 [Bug Report] Blackblade of Shahram - Guardian is not casting his spells

### ISSUE THREAD

**Issue by fatalieror** [cmangos/issues#2568]
### Expected behavior

https://www.youtube.com/watch?v=YGhVEasWOzU
https://www.youtube.com/watch?v=tZYPyCeeFpw

MEELE HIT procs the ONHIT chance of Blackblade and summons Shahram (10718)

**Sharam randomly casts 1 of the following spells and disappears, which also means his Duration Time has to be reduced to ~7 seconds.**

1. https://classicdb.ch/?spell=16599 Blessing of Shahram
2. https://classicdb.ch/?spell=16601 Fist of Shahram
3. https://classicdb.ch/?spell=16598 Will of Shahram
4. https://classicdb.ch/?spell=16597 Curse of Shahram
5. https://classicdb.ch/?spell=16596 Flames of Shahram
6. https://classicdb.ch/?spell=16600 Might of Shahram

I Managed to make him cast the spells randomly by grouping his buffs and offensive spells together.
But now he casts both groups : 1 offensive 1 buff. though he is supposed to cast only 1 spell and disappear.

Also in the vidoes it seems, that Shahram is passive and stationary. Right now my last "ActionTyp" to put him in ReactState_Passive doesnt seem to work. 

Would apreciate some pointers regarding the ReactState, because that was the hardest to find the wiki. Also how do you make the randomness check beforehand of the actual cast?

See below: Creature_AI_scripts

`  id  creature_id  event_type  event_inverse_phase_mask  event_chance  event_flags  event_param1  event_param2  event_param3  event_param4  event_param5  event_param6  action1_type  action1_param1  action1_param2  action1_param3  action2_type  action2_param1  action2_param2  action2_param3  action3_type  action3_param1  action3_param2  action3_param3  comment                                
-------  -----------  ----------  ------------------------  ------------  -----------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  --------------  --------------  --------------  ------------  --------------  --------------  --------------  ------------  --------------  --------------  --------------  ---------------------------------------
1071801        10718          11                         0           100            0             0             0             0             0             0             0            57               2              35               0             0               0               0               0             0               0               0               0  Shahram - Enable Range Mode on Aggro   
1071802        10718           0                         0           100         1056             0            40          3400          4800             0             0            11           16600               1               0            11           16597               1               0            11           16596               1               0  Shahram - Might of Shahram             
1071803        10718          11                         0           100         1056             0             0             1             1             0             0            11           16601              11               0            11           16598              11               0            11           16599              11               0  Shahram - Fist of Shahram Group Haste  
1071806        10718          11                         0             0         1025             0             0             0             0             0             0            50               0               0               0            57               2              40               0             0               0               0               0  Shahram - ReactState Passive           `

### Version & Environment

  - "1.12.1" (CLASSIC)
commit a4cccb23a3734ddc10e5827220b77c0da5dd90ca
[  - [CLASSIC](https://github.com/cmangos/classic-db/tree/XXXX)](https://github.com/cmangos/classic-db)

  - Win XX

### Steps to reproduce
1. Equip Weapon "Blackblade of Shahram
2. Hit a Mob
3. Observe how Shahram is Spawned but he does nothing but Auto Attack
4. Profit

edit anonxs: use s.16606 to scrip this.

**Comment by anythingll** [comment]
What you need sir, is phases, make him pick a random phase between 1-6 (action 31) at spawn and set a line and a phasemask corresponding to 1-6 for each of his spells
First render him unkillable/untargetable with this

UPDATE `creature_template` SET `UnitFlags`= UnitFlags
|2 -- UNIT_FLAG_NON_ATTACKABLE
|33554432 -- UNIT_FLAG_NOT_SELECTABLE
WHERE `entry` = 10718; -- Shahram

Then for the ai_scripts

INSERT INTO `creature_ai_scripts` (`id`, `creature_id`, `event_type`, `event_inverse_phase_mask`, `event_chance`, `event_flags`, `event_param1`, `event_param2`, `event_param3`, `event_param4`, `action1_type`, `action1_param1`, `action1_param2`, `action1_param3`, `action2_type`, `action2_param1`, `action2_param2`, `action2_param3`, `action3_type`, `action3_param1`, `action3_param2`, `action3_param3`, `comment`) VALUES
(1071801, 10718, 11, 0, 100, 0, 0, 0, 0, 0, 50, 0, 0, 0, 31, 1, 6, 0, 11, 17507, 0, 2, 'Shahram - Set ReactStates:0, Random phase and Root on Spawn'),
(1071802, 10718, 29, 124, 100, 0, 1000, 1000, 0, 0, 11, 16596, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 'Shahram - Cast Flames of Shahram (P1)'),
(1071803, 10718, 29, 122, 100, 0, 1000, 1000, 0, 0, 11, 16597, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 'Shahram - Cast Curse of Shahram (P2)'),
(1071804, 10718, 29, 118, 100, 0, 1000, 1000, 0, 0, 11, 16598, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 'Shahram - Cast Will of Shahram (P3)'),
(1071805, 10718, 29, 110, 100, 0, 1000, 1000, 0, 0, 11, 16599, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 'Shahram - Cast Blessing of Shahram (P4)'),
(1071806, 10718, 29, 94, 100, 0, 1000, 1000, 0, 0, 11, 16600, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 'Shahram - Cast Might of Shahram (P5)'),
(1071807, 10718, 29, 62, 100, 0, 1000, 1000, 0, 0, 11, 16601, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 'Shahram - Cast Fist of Shahram (P6)'),
(1071808, 10718, 7, 0, 100, 0, 0, 0, 0, 0, 11, 17507, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 'Shahram - Root on Evade'),
(1071809, 10718, 29, 0, 100, 0, 7000, 7000, 0, 0, 41, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 'Shahram - Despawn on Generic Timer');

This is as close as I can reproduce his behavior

**Comment by killerwife** [comment]
@AnonXS check this EAI code out.

**Comment by anythingll** [comment]
Mobs remove their buff/debuff when they leave combat and since he roots himself at spawn you dont want him to wander around during that short period of time (there may be better way to lock mobs in place but I havent figured yet)

**Comment by killerwife** [comment]
There is a spell_template custom attribute for that evade aura remove actually.

**Comment by AnonXS** [comment]
First: Spell is missing in tbc+, needs to be ported and second: Dont we use IsSpellRemovedOnEvade for this?

It has the typical attribute making it an aura passive, which is not removed on evade:
```
SpellAttributes

SPELL_ATTR_PASSIVE - 64
SPELL_ATTR_DO_NOT_DISPLAY - 128
SPELL_ATTR_HIDE_IN_COMBAT_LOG - 256

AttributesEx

SPELL_ATTR_EX_DONT_DISPLAY_IN_AURA_BAR
```

**Comment by killerwife** [comment]
I added SPELL_ATTR_SS_IGNORE_EVADE so that the evergrowing list in c++ can be moved to DB.

**Comment by al3xc1985** [comment]
This might help
c mangos introduced it I believe
https://github.com/vmangos/core/commit/2c4ec612dcedb33c4fd3903ac221a67e89932eef
https://github.com/vmangos/core/commit/2c4ec612dcedb33c4fd3903ac221a67e89932eef

**Comment by MantisLord** [comment]
I started with @anythingll's idea, I just condensed it to two phases and used EFLAG_RANDOM_ACTION to random between 3 spells per phase. You could also use a single ACTION_T_START_RELAY_SCRIPT to launch a dbscript random template with 6 possible outcomes.
You were on the right track I think by using passive react state. I did the same thing, but also used ACTION_T_SET_FOLLOW_MOVEMENT to turn his follow movement off instead of the root spell and that seemed to do the trick.

> First render him unkillable/untargetable with this
> 
> UPDATE `creature_template` SET `UnitFlags`= UnitFlags
> |2 -- UNIT_FLAG_NON_ATTACKABLE
> |33554432 -- UNIT_FLAG_NOT_SELECTABLE
> WHERE `entry` = 10718; -- Shahram

I don't have him sniffed, so I'm a bit on the fence about adding the unselectable unitflag without some kind of evidence. Can anyone confirm you can't select them on classic?

**Comment by anythingll** [comment]
Yeah I ended up doing the same with the EFLAG_RANDOM_ACTION, takes way less space, forgot to update this post tho.
Can't tell for sure about the NOT_SELECTABLE flag but this video https://youtu.be/YGhVEasWOzU?si=M7WhRlEq9gQ4KNJ0&t=110 shows that the guy has allied pets/guardians name enabled since we can see the allied succubus name and yet Shahram name/title isnt displayed, which is one of the NOT_SELECTABLE flag feature.

**Comment by al3xc1985** [comment]
https://youtu.be/YGhVEasWOzU?t=63
Doesn't this mean that if nova does not affect them (images of sha...) thet are unselectable?

Check the video at that point
Dunno, just asking

**Comment by anythingll** [comment]
Yeah looks like you cannot harm Shahram, he will cast his thing no matter what

**Comment by AnonXS** [comment]
Needs: CREATURE_EXTRA_FLAG_FORCE_ATTACKING_CAPABILITY?
